openhtf/conf.py

```diff
@@ -43,250 +128,224 @@
-        and not declaration.optional):
-      raise MissingRequiredConfigurationKeyError(
-          declaration.name, declaration.description)
+    self._flags = flags
+    self._logger = logger
+    self._lock = lock
+    self._functools = _functools
+    self._modules = kwargs
+    self._declarations = {}
+    self._loaded_values = {}
+    self._flag_values = {}
+    for keyval in self._flags['config-value'].value:
+      self._flag_values.setdefault(*keyval.split('=', 1))
+
+    # Everywhere that uses configuration uses this, so we just declare it here.
+    self.Declare('station_id', 'The name of this tester')
+
+  # Don't use Synchronized on this one, because __getitem__ handles it.
+  def __getattr__(self, attr):  # pylint: disable=invalid-name
+    """Get a config value via attribute access."""
+    if attr and attr[0].islower():
+      return self[attr]
+    # Config keys all begin with a lowercase letter, so treat this normally.
+    raise AttributeError("'%s' object has no attribute '%s'" %
+                         (type(self).__name__, attr))
 
-    if value is None:
-      return declaration.default_value
-    return value
-
-  def __contains__(self, name):  # pylint: disable=invalid-name
-    return name in self._declared
-
-  def __getitem__(self, name):  # pylint: disable=invalid-name
-    return self._declared[name]
-
-  def __copy__(self):  # pylint: disable=invalid-name
-    self_copy = type(self)()
-    for name, declaration in self._declared.iteritems():
-      self_copy.Declare(name, declaration)
-    return self_copy
-
-
-class ConfigModel(object):
-  """A model that holds the underlying config keys and their values.
-
-  By isolating the underlying model it provides a way to lock access to the
-  dictionary so we can reload it on demand or otherwise poke it.
-  """
+  @threads.Synchronized
+  def __getitem__(self, item):  # pylint: disable=invalid-name
+    """Get a config value via item access.
 
-  def __init__(self, state=None, declarations=None):
-    """Initializes the model.
+    Order of precedence is:
+      - Value provided via --config-value flag.
+      - Value loaded via Load*() methods.
+      - Default value as declared with conf.Declare()
 
     Args:
-      state: A dictionary containing configuration key, values.  By default a
-          new one is created.  If one is provided the model is marked as
-          loaded.
-      declarations: An object which tracks declared keys, if not provided
-          a new one is constructed.
+      item: Config key name to get.
     """
-    self._state = state if state is not None else {}
-    self._declarations = declarations or _DeclaredKeys()
-    self._loaded = state is not None
-    self.lock = threading.Lock()
+    if item not in self._declarations:
+      raise self.UndeclaredKeyError('Configuration key not declared', item)
+
+    if item in self._flag_values:
+      if item in self._loaded_values:
+        self._logger.warning(
```

Comment:
Ah, I see what you meant.  Yeah I like keeping them around separately because it does allow me to treat flag-provided values as second class citizens and badger the user for using them, and it also lets me distinguish between overriding a `default_value` (which shouldn't log a message, that's an obvious thing to do) and overriding a previously `Load()`ed value.

A snapshot view of the config is available via `_asdict()`, incidentally, but I'm just not using it to detect deltas like you're talking about.